fix: ignore pip version warnings, is_installed error handling

n-dusan opened the pull request:
* If pip version is not upgraded, pip shows a pip version WARNING which messes up
  the json parse.  This is happens because `run` result output also returns errors and warnings, which we use to determine errors in other parts of our platform. 
* Use `JSONDecoder` to parse a string into `json`, ignoring extra string contents at new lines. 
* After parsing `pip list` into JSON, return None if version of package does not exist. In which case we log the result.
* Update script except handling, which should raise an error and make script fail. 
* Legacy pip (for example `9.0.3`) does not have `config` method. This can cause an error in the script, which shouldn't be the case. We just wish to append `--pre` to development packages, so a `--pre` flag can be sent from the script nargs.

## Steps to reproduce issue
* [x] In a new virtual environment, install latest upgrade-python-package script from pypi (`==0.3.0`)
* [x] Downgrade pip to an older version, (if wish to go really old, try `pip 9.0.3`)
* [x] Run upgrade-python-package script,  (example: `python -m upgrade.scripts.upgrade_python_package requests --test`. Expect `json.decoder.JSONDecodeError` in `is_package_already_installed`
* [x] Install current upgrade-python-package script in editable mode
* [x] Run upgrade-python-package script, expect everything working as expected

joelbarker2011 commented:
Some concerns:
1. I would like to see an example in the PR checklist that shows the before/after behavior (and demonstrates that this will fix the reported bug).
2. If I'm not mistaken, the versions returned by `pip list` will be surrounded by parentheses, which will always fail the `==` comparison. (Not tested.)
3. Catching and ignoring all possible exceptions is generally bad practice. We have no idea what exceptions might be expected to arise, and we're just silently ignoring them anyway. Can we restrict to some expected exception? If we don't know what to expect, why would we ignore exceptions here? At a minimum, shouldn't we at least be logging?
4. By removing `--format json`, we're eliminating the only guarantee we can possibly get about the output of `pip list`. Unless the `pip` documentation says otherwise (I haven't checked), we can't reliably count on the output of `pip list` (with no format) to be consistent or stable. So this adds a ticking time bomb to our code.
5. It seems to me that the true underlying bug is the fact that the `run` method redirects `stderr` to `stdout`, thus polluting the JSON output of the `pip` command with the pip upgrade warning. This should be refactored IMO (with due care for any code that might rely on this behavior). See https://github.com/openlawlibrary/upgrade-python-package/blob/678c9265a9803c8706b918f4f50de76fbb56718b/upgrade/scripts/upgrade_python_package.py#L350

n-dusan committed:
fix: use JSONDecoder to parse pip list in json format

* If there's trailing output in pip list --format json, response will
 contain cluttered strings which can't be decoded into json. Use
 JSONDecoder to parse string into json, and an index at which point JSON
 can't be converted anymore. Log warn if package isn't found.

n-dusan committed:
feat: unit test `is_package_already_installed`

n-dusan committed:
fix: lazy load requests module

*If upgrade python package script is installed without a requests module, lazy load the import into the function. This can happen if script is being used by a very old `setuptools`<50 version

n-dusan committed:
fix: silently continue if `config` isn't found

*When using legacy pip, `config` command does not exist. In that case, we want to append '--pre' to development (cloudsmith) repositories, which can be passed down as an upgrade script narg. Example: python -m upgrade.scripts.upgrade_python_package oll-* '--pre'

n-dusan committed:
fix: let script fail if error is raised and output as json is false

n-dusan commented:
@joelbarker2011  I agree [1-4]. We revert back to `--format json` approach but with added `json.JSONDecoder` which can parse a string correctly. Error handling seems a bit better now.
About the underlying issue suspicions, after going over with it @renatav, I don't believe we need to refactor?

renatav commented:
We have some code similar to the previous solution in another place, though there we handle editable wheels separately. That's a good point about not relying on the default format staying the default one, we should probably update that code with `--format columns`. I would not expect any errors to occur while executing `pip list`, the except clause was there instead of checks which would ensure that a line consists of a package name version pair (thus skipping warnings and errors). I've seen error handling being used in this way quite a few times and I find it to be intuitive enough, but we could've caught a more specific error. We could've also added checks which would've prevented any errors from being raise in the first place. We probably don't need to support very old pip versions (considering out use cases), so instead of having additional code which would graciously handle editable wheels and errors, I do vote for going back to the json format and using the decoder to extract the json part.

About the underlying issue, that sounds like a reasonable solution, but we've been using this same code since the early stages of the project, so we are used to dealing with error messages and rely on them in quite a few places, so not sure that we want to focus on that rework now. Should be easy enough to disregard the errors regardless of which format we end up using.

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -216,7 +220,12 @@ def is_cloudsmith_url_valid(cloudsmith_url):
 def is_package_already_installed(package):
     results = pip("list", "--format", "json")
-    parsed_results = json.loads(results)
+    try:
```
After all of these discussions and worries about format of pip output, errors, pollution of the script execution output, I'm thinking if we can maybe avoid using pip altogether. Maybe `importlib`? https://blog.finxter.com/python-check-version-of-package-with-pip
Though we really should not spend too much more time and effort into addressing some rather unlikely scenarios imo.
Might even be better to revert this implementation to `git show`, despite that shortcoming in order to avoid dropping support for older pip. But again, not sure if we even care about pip version which are so old

n-dusan replied:
That looks cool and could potentially be a simpler solution. But as I've looked through `importlib`, it seems that they are new in >=3.8. Docs found here https://docs.python.org/3.8/library/importlib.metadata.html
In 3.6 it raises an error.
```
ModuleNotFoundError: No module named 'importlib.metadata'
```
So I think we'll have to keep going with this approach for a little while into the future, until we completely stop supporting 3.6

renatav replied:
Right, amid Python 3.10 reworks, I forgot that we still need this script to support Python 3.6. Can you also revert the drop 3.6 support changes in this PR?

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -216,7 +220,12 @@ def is_cloudsmith_url_valid(cloudsmith_url):
+        parsed_results, index = decoder.raw_decode(results)
+    except json.JSONDecodeError as e:
+        logging.error(f"Error occurred while decoding pip list at ${index} index.")
+        raise e
```
Do we want the script to fail if this fails? Maybe we do in order to avoid any chance of having broken dependencies. However, in that case we might want to ask the user to upgrade pip. So this can be a custom exception.

n-dusan replied:
We probably want to fail if pip --format json isn't parsed into a json correctly. That would IMO indicate that a deprecated (or potentially newer, breaking) pip is used? I'll add a custom exception

renatav replied:
The question is if we want to support really old pio too, or make sure that, say, version 18 is new enough.

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -226,6 +235,7 @@ def is_package_already_installed(package):
     if found_package:
         _, version = found_package.pop()
         return version
+    logging.warning(f"Package not found: ${package}")
```
Think this should just be `info`

renatav reviewed (commented):
Feel like I've gone crazy from overthinking issues with `pip` versions now, but here are some comments

n-dusan committed:
fix: json decode error handling

* Create a custom exception which asserts that pip list --format json
  to json conversion failed. This can be a useful message to user to
  upgrade their pip.
* info instead of warning logging

feat: make script fail if pip list to json fails

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -216,7 +222,13 @@ def is_cloudsmith_url_valid(cloudsmith_url):
+    except json.JSONDecodeError as e:
+        msg = f"Error occurred while decoding pip list at ${index} index."
+        logging.error(msg)
+        raise PipFormatDecodeFailed(msg)
```
We could catch this exception in `install_wheel` and print a message saying that it might be necessary to upgrade pip

n-dusan committed:
fix: bring back 3.6

fix: 3.6 compatible cloudsmith test

fix: pin setuptools_scm to 6.*

joelbarker2011 commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -216,7 +230,13 @@ def is_cloudsmith_url_valid(cloudsmith_url):
+        decoder = json.JSONDecoder()
+        parsed_results, index = decoder.raw_decode(results)
+    except json.JSONDecodeError as e:
+        msg = f"Error occurred while decoding pip list at ${index} index."
```
`index` will not be defined here if an exception was thrown on line 235.

n-dusan replied:
My bad

joelbarker2011 reviewed (changes requested):
@n-dusan, testing worked for me, although I had to add a line to verify that the old `main` code was raising an exception:
```diff
      except Exception as e:
+         print("#####", type(e).__name__, "#####")
```

Requesting 1 change due to an unrelated error that I noticed.

n-dusan committed:
fix: `index` is not defined in exception

joelbarker2011 reviewed (approved):
Changes LGTM. I have no knowledge of some of these changes, but I see no problems and @renatav approved them separately.
